File: forex-server/app/request/CandlesDownloadRequest.scala
package request

import play.api.libs.json.{JsObject, JsResult, JsValue, Json, OFormat}

/** Case class used to represent user requests for downloading historical candle
  * data.
  *
  * @param currencyPair
  *   The currency pair to download data for.
  * @param granularity
  *   The granularity to retrieve (minute candles, 30-minute candles, hourly
  *   candles, etc.)
  * @param pricingComponents
  *   Which prices to include in the candles (bid, mid, and/or ask).
  * @param fromDate
  *   The start datetime to download data for. Should be in yyyy-MM-dd HH:mm:ss"
  *   format.
  * @param toDate
  *   The end datetime to download data for. Should be in yyyy-MM-dd HH:mm:ss"
  *   format.
  */
case class CandlesDownloadRequest(
    currencyPair: CurrencyPair,
    granularity: Granularity,
    pricingComponents: Seq[PricingComponent],
    fromDate: String,
    toDate: String
)

/** Companion CandlesDownloadRequest object used for defining JSON
  * functionality.
  */
object CandlesDownloadRequest {
  implicit val candlesDownloadRequestFormat: OFormat[CandlesDownloadRequest] =
    Json.format[CandlesDownloadRequest]
}

File: forex-frontend/src/app/downloader/request/candles-download-request.ts
import { CurrencyPair } from "./currency-pair.enums";
import { Granularity } from "./granularity.enums";
import { PricingComponent } from "./pricing-component.enums";

/**
 * Simple class that holds data for a candles download request.
 */
export class CandlesDownloadRequest {
    /**
     * Class constructor.  Nothing special is needed since we just use the class
     * to store data.
     *
     * @param currencyPair Currency pair to download data for.
     * @param granularity Time frame/granularity of the data that the user wants.
     * @param pricingComponents Which price components to include (bid, mid, and/or ask).
     * @param fromDate Starting date to download data.
     * @param toDate End/final date to download data.
     */
    constructor(
        public currencyPair: CurrencyPair,
        public granularity: Granularity,
        public pricingComponents: PricingComponent[],
        public fromDate: string,
        public toDate: string
    ) {}
}
